File: scala/robot-name/src/main/scala/robot_name.scala
import scala.annotation.tailrec

class Robot {

  private var generatedName = Robot.nextUniqueName

  def name(): String = generatedName

  def reset(): Unit = {
    generatedName = Robot.nextUniqueName
  }

}

object Robot {

  private var usedNames = scala.collection.mutable.Set.empty
  private val names: Stream[String] = RandomNames().stream
  
  def nextUniqueName(): String = {
    names.take(1).mkString
  }

}

case class RandomNames() {

  private val letters = new RandomBag('A' to 'Z').stream
  private val digits = new RandomBag(0 to 9).stream

  def generateName(): String = {
    letters.take(2).mkString ++ digits.take(3).mkString
  }

  def stream: Stream[String] = {
    Stream.continually(generateName)
  }

}

class RandomBag[T](items: IndexedSeq[T]) {

  private val numbers = new scala.util.Random
  private val range = items.size  

  def randomIndex = numbers.nextInt(range)

  def stream: Stream[T] = {
    Stream.continually(items(randomIndex))
  }
  
}

File: ruby/robot-name/robot_name.rb
require 'set'

class RandomNames

  attr_reader :letters, :digits

  def initialize
    @letters = 'A'.upto('Z').to_a
    @digits = 0.upto(9).to_a
  end

  def next
    generate_name
  end

  def generate_name
    (take_random(2, letters) + take_random(3, digits)).join
  end

  def take_random(num, items)
    1.upto(num).map { |i| items.sample }
  end

end

class UniqueStream

  def initialize(stream)
    @used = Set.new
    @stream = stream
  end

  def next
    loop do
      item = @stream.next

      return item if @used.add?(item)
    end
  end

end

class Robot

  attr_reader :name

  @@names = UniqueStream.new(RandomNames.new)

  def initialize
    reset
  end

  def reset
    @name = @@names.next
  end

end

module BookKeeping
  VERSION = 2
end
